Refactor/5-app-infra step (#45)
GoogleCloudPlatform/terraform-google-enterprise-genai: 5-app-infra/README.md

</table>

For an overview of the architecture and the parts, see the
[terraform-google-enterprise-genai README](https://github.com/GoogleCloudPlatform/terraform-google-enterprise-genai)
file.

## Purpose

> Inside the `projects` folder, the folders `artifact-publish` and `service-catalog` are projects that will be _expanded_ upon.  In step 4, we have initiated the creation of these projects, enabled API's and assigned roles to various service accounts, service agents and cryptography keys that are needed for each respective project to operate successfully. Inside the `source_repos` folder, the folders `artifact-publish` and `service-catalog` are seperate cloud build repositories that have their own unique piplelines configured. These are used for building out in-house Docker images for your machine-learning pipelines and terraform modules that will be used in `notebooks` in your interactive (development) environment, as well as deployment modules for your operational (non-production, production) environments respectively.

Reviewer: ```suggestion
Inside the `projects` folder, the `artifact-publish` and `service-catalog` directories contain applications that will be further developed. These directories are Terraform repositories that house the configuration code for their respective applications. For instance, in the `projects/artifact-publish` directory, you will find code that configures the custom pipeline for the artifact-publish application.
> Note: Remember that in step 4-projects, the Service Catalog and Artifacts projects were created under `common` folder.

Inside the `source_repos` folder, the folders `artifact-publish` and `service-catalog` are seperate Cloud Build Repositories that have their own unique piplelines configured. These are used for building out in-house Docker images for your machine-learning pipelines and terraform modules that can be deployed through the Service Catalog Google Cloud Product.

This repository contain examples using modules in `notebooks` in your interactive (development) environment, as well as deployment modules for your operational (non-production, production) environments respectively.
```

For the purposes of this demonstration, we assume that you are using Cloud Build or manual deployment.

## Prerequisites

> 1. 0-bootstrap executed successfully.

Reviewer: Fix order of list here, from 1-5

1. 1-org executed successfully.
1. 2-environments executed successfully.
1. 3-networks executed successfully.
1. 4-projects executed successfully.

### Troubleshooting

Please refer to [troubleshooting](../docs/TROUBLESHOOTING.md) if you run into issues during this step.

## Usage

**Note:** If you are using MacOS, replace `cp -RT` with `cp -R` in the relevant
commands. The `-T` flag is needed for Linux, but causes problems for MacOS.

### Deploying with Cloud Build

> 

Reviewer: Also fix order of list here

1. Ensure you are in a neutral directory outside any other git related repositories.

> 1. Clone the `gcp-policies` repo based on the Terraform output from the `0-bootstrap` step.

Reviewer: ```suggestion
1. Clone the `gcp-policies` repo based on the Terraform output from the `4-projects` step.
```

   ```

1. `cd` out of the `publish-artifacts` repository.
   
   ```bash
   cd ..
   ```

1. Grab the Service Catalogs ID
  
   ```bash
   export SERVICE_CATALOG_PROJECT_ID=$(terraform -chdir="gcp-projects/business_unit_3/shared" output -raw service_catalog_project_id)
   echo ${SERVICE_CATALOG_PROJECT_ID}
   ```

1. Clone the freshly minted Cloud Source Repository that was created for this project.
   
   ```bash
   gcloud source repos clone service-catalog --project=${SERVICE_CATALOG_PROJECT_ID}
   ```

1. Enter the repo folder and copy over the service catalogs files from `5-app-infra/source_repos/service-catalog` folder.
   
   ```bash
   cd service-catalog/
   cp -RT ../terraform-google-enterprise-genai/5-app-infra/source_repos/service-catalog/ .
   ```

1. Commit changes and push main branch to the new repo.
   
   ```bash
   git add .
   git commit -m 'Initialize Service Catalog Build Repo'

   git push --set-upstream origin main
   ```

1. `cd` out of the `service_catalog` repository.
   
   ```bash
   cd ..
   ```

1. Navigate to the project that was output from `${ARTIFACT_PROJECT_ID}` in Google's Cloud Console to view the first run of images being built.

> #### VPC-SC

Reviewer: Remove VPC-SC and INGRESS/EGRESS policies

Move it to the 6-ml-pipeline

Also note that on repo root contains [6-machine-learning](https://github.com/GoogleCloudPlatform/terraform-google-enterprise-genai/tree/a39cce24cc129f6e79c2b1c7953a38a5c91f17d5/6-machine-learning) and [6-ml-pipeline/dev](https://github.com/GoogleCloudPlatform/terraform-google-enterprise-genai/tree/a39cce24cc129f6e79c2b1c7953a38a5c91f17d5/6-ml-pipeline/dev)

both should be merged into one

Since every workflow will have access to these images, it is deployed in the `common` folder, and keeping with the foundations structure, is listed as `shared` under this Business Unit.  It will only need to be deployed once.

The Pipeline is connected to a Google Cloud Source Repository with a simple structure:

   ```
   ├── README.md
   └── images
      ├── tf2-cpu.2-13:0.1
      │   └── Dockerfile
      └── tf2-gpu.2-13:0.1
         └── Dockerfile
   ```
for the purposes of this example, the pipeline is configured to monitor the `main` branch of this repository.

each folder under `images` has the full name and tag of the image that must be built.  Once a change to the `main` branch is pushed, the pipeline will analyse which files have changed and build that image out and place it in the artifact repository.  For example, if there is a change to the Dockerfile in the `tf2-cpu-13:0.1` folder, or if the folder itself has been renamed, it will build out an image and tag it based on the folder name that the Dockerfile has been housed in.

Once pushed, the pipeline build logs can be accessed by navigating to the artifacts project name created in step-4:

   ```bash

>    terraform -chdir="terraform-example-foundation/4-projects/business_unit_3/shared/" output -raw common_artifacts_project_id

Reviewer: ```suggestion
   terraform -chdir="gcp-projects/business_unit_3/shared/" output -raw common_artifacts_project_id
```

   echo "backend_bucket = ${backend_bucket}"

   for i in `find -name 'backend.tf'`; do sed -i "s/UPDATE_APP_INFRA_BUCKET/${backend_bucket}/" $i; done
   ```

6. Commit changes.

   ```bash
   git add .
   git commit -m 'Initialize repo'
   ```

7. Push your plan branch to trigger a plan for all environments. Because the
   _plan_ branch is not a [named environment branch](../docs/FAQ.md#what-is-a-named-branch), pushing your _plan_
   branch triggers _terraform plan_ but not _terraform apply_. Review the plan output in your Cloud Build project https://console.cloud.google.com/cloud-build/builds;region=DEFAULT_REGION?project=YOUR_INFRA_PIPELINE_PROJECT_ID

   ```bash
   git push --set-upstream origin plan
   ```

8. Merge changes to shared. Because this is a [named environment branch](../docs/FAQ.md#what-is-a-named-branch),
   pushing to this branch triggers both _terraform plan_ and _terraform apply_. Review the apply output in your Cloud Build project https://console.cloud.google.com/cloud-build/builds;region=DEFAULT_REGION?project=YOUR_INFRA_PIPELINE_PROJECT_ID

   ```bash
   git checkout -b shared

>    git push origin shared

Reviewer: ```suggestion
   git push origin production
```

   ```

9. `cd` out of the `bu3-artifacts-publish` repository.

   ```bash
   cd ..
   ```

> #### Configuring Cloud Source Repositorie of Artifact Application

Reviewer: ```suggestion
#### Configuring Cloud Source Repository of Artifact Application
```

   echo "backend_bucket = ${backend_bucket}"

   for i in `find -name 'backend.tf'`; do sed -i "s/UPDATE_APP_INFRA_BUCKET/${backend_bucket}/" $i; done
   ```

6. Commit changes.

   ```bash
   git add .
   git commit -m 'Initialize repo'
   ```

7. Push your plan branch to trigger a plan for all environments. Because the
   _plan_ branch is not a [named environment branch](../docs/FAQ.md#what-is-a-named-branch), pushing your _plan_
   branch triggers _terraform plan_ but not _terraform apply_. Review the plan output in your Cloud Build project https://console.cloud.google.com/cloud-build/builds;region=DEFAULT_REGION?project=YOUR_INFRA_PIPELINE_PROJECT_ID

   ```bash
   git push --set-upstream origin plan
   ```

8. Merge changes to shared. Because this is a [named environment branch](../docs/FAQ.md#what-is-a-named-branch),
   pushing to this branch triggers both _terraform plan_ and _terraform apply_. Review the apply output in your Cloud Build project https://console.cloud.google.com/cloud-build/builds;region=DEFAULT_REGION?project=YOUR_INFRA_PIPELINE_PROJECT_ID

   ```bash

>    git checkout -b shared

Reviewer: ```suggestion
   git checkout -b production
```

>    git push origin shared

Reviewer: ```suggestion
   git push origin production
```

   ```

9. `cd` out of the `bu3-service-catalog` repository.
   
   ```bash
   cd ..
   ```

> #### Configuring Cloud Source Repositorie of Service Catalog Solutions Artifacts

Reviewer: ```suggestion
#### Configuring Cloud Source Repository of Service Catalog Solutions Pipeline
```

1. Grab the Service Catalogs ID

   ```bash
   export SERVICE_CATALOG_PROJECT_ID=$(terraform -chdir="gcp-projects/business_unit_3/shared" output -raw service_catalog_project_id)
   echo ${SERVICE_CATALOG_PROJECT_ID}
   ```

2. Clone the freshly minted Cloud Source Repository that was created for this project.
   
   ```bash
   gcloud source repos clone service-catalog --project=${SERVICE_CATALOG_PROJECT_ID}
   ```

3. Enter the repo folder and copy over the service catalogs files from `5-app-infra/source_repos/service-catalog` folder.
   
   ```bash
   cd service-catalog/

>    cp -RT ../terraform-google-enterprise-genai/5-app-infra/source_repos/service-catalog/ .

Reviewer: ```suggestion
   git commit -m "Initialize Repository" --allow-empty
   cp -RT ../terraform-google-enterprise-genai/5-app-infra/source_repos/service-catalog/ .
```

   ```

4. Commit changes and push main branch to the new repo.
   
   ```bash

>    git add .

Reviewer: ```suggestion
   git add modules
```

   git commit -m 'Initialize Service Catalog Build Repo'

   git push --set-upstream origin main
   ```

5. `cd` out of the `service_catalog` repository.
   
   ```bash
   cd ..
   ```

6. Navigate to the project that was output from `${ARTIFACT_PROJECT_ID}` in Google's Cloud Console to view the first run of images being built.

> ### Run Terraform locally

Reviewer: Run Terraform locally section is outdated

You have two options:
- Remove it entirely and assume the user will always run cloud build
- Update it with the missing steps
